service/backend/pchain/construction.go

```diff
@@ -14,12 +17,154 @@ func (b *Backend) ConstructionDerive(ctx context.Context, req *types.Constructio
+	if err != nil {
+		return nil, service.WrapError(service.ErrInvalidInput, err)
+	}
+
+	metadata.BlockchainID = pChainID
+
+	metadataMap, err := mapper.MarshalJSONMap(metadata)
+	if err != nil {
+		return nil, service.WrapError(service.ErrInternalError, err)
+	}
+
+	return &types.ConstructionMetadataResponse{
+		Metadata:     metadataMap,
+		SuggestedFee: []*types.Amount{suggestedFee},
+	}, nil
+}
+
+func (b *Backend) buildImportMetadata(ctx context.Context, options map[string]interface{}) (*pmapper.Metadata, *types.Amount, error) {
+	var preprocessOptions pmapper.ImportExportOptions
+	if err := mapper.UnmarshalJSONMap(options, &preprocessOptions); err != nil {
+		return nil, nil, err
+	}
+
+	sourceChainID, err := b.pClient.GetBlockchainID(ctx, preprocessOptions.SourceChain)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	suggestedFee, err := b.getBaseTxFee(ctx)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	importMetadata := &pmapper.ImportMetadata{
+		SourceChainID: sourceChainID,
+	}
+
+	return &pmapper.Metadata{ImportMetadata: importMetadata}, suggestedFee, nil
+}
+
+func (b *Backend) buildExportMetadata(ctx context.Context, options map[string]interface{}) (*pmapper.Metadata, *types.Amount, error) {
+	var preprocessOptions pmapper.ImportExportOptions
+	if err := mapper.UnmarshalJSONMap(options, &preprocessOptions); err != nil {
+		return nil, nil, err
+	}
+
+	destinationChainID, err := b.pClient.GetBlockchainID(ctx, preprocessOptions.DestinationChain)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	suggestedFee, err := b.getBaseTxFee(ctx)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	exportMetadata := &pmapper.ExportMetadata{
+		DestinationChain:   preprocessOptions.DestinationChain,
+		DestinationChainID: destinationChainID,
+	}
+
+	return &pmapper.Metadata{ExportMetadata: exportMetadata}, suggestedFee, nil
+}
+
+func (b *Backend) buildStakingMetadata(options map[string]interface{}) (*pmapper.Metadata, *types.Amount, error) {
+	var preprocessOptions pmapper.StakingOptions
+	if err := mapper.UnmarshalJSONMap(options, &preprocessOptions); err != nil {
+		return nil, nil, err
+	}
+
+	stakingMetadata := &pmapper.StakingMetadata{
+		NodeID:          preprocessOptions.NodeID,
+		Start:           preprocessOptions.Start,
+		End:             preprocessOptions.End,
+		Wght:            preprocessOptions.Wght,
+		Memo:            preprocessOptions.Memo,
+		Locktime:        preprocessOptions.Locktime,
+		Threshold:       preprocessOptions.Threshold,
+		RewardAddresses: preprocessOptions.RewardAddresses,
+		Shares:          preprocessOptions.Shares,
+	}
+
+	zeroAvax := mapper.AvaxAmount(big.NewInt(0))
+
+	return &pmapper.Metadata{StakingMetadata: stakingMetadata}, zeroAvax, nil
+}
+
+func (b *Backend) getBaseTxFee(ctx context.Context) (*types.Amount, error) {
+	fees, err := b.pClient.GetTxFee(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	feeAmount := new(big.Int).SetUint64(uint64(fees.TxFee))
```

Comment:
The fees returned here are in `nAVAX` that has decimals 9. We may need a separate `mapper.AvaxAmount` for the P chain.

Reply:
Introduced the `AtomicAvaxAmount` function and a corresponding currency with 9 decimals